### ISSUE TITLE
New item

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8486]
**Describe the Bug**

Recently we saw in production when a practitioner associated with a large number of cases updated their contact information the lambda responsible for issuing applicable Notices of Change (in this case the Notice of Change of Address) timed out prior to docketing all the expected notices.

After discussion in standup, it was suggested that the mechanism for this process could be refactored to ensure completion of all appropriate updates and avoid long-running operations timing out (much the same way we did when generating calendaring notices was failing to complete).

The most recent example we have is practitioner bar# AD0238. - an update to the address prompted the usual generation of Notice of Change of Address, however the lambda timed out before the operation could complete and subsequently not all notices were generated/docketed and not all service emails were sent. 

**Business Impact/Reason for Severity**
Especially with practitioners who regularly appear before Tax Court, the list of cases that must be updated when contact info is changed can be incredibly long (particularly with IRS practitioners). We need to ensure reliability for this operation to prevent the need for manual remedies.

**In which environment did you see this bug?**
PROD

**Who were you logged in as?**

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**

**To Reproduce**
1. Log in as a petitioner or practitioner with a large number of cases
2. Update and save change(s) to your contact information.

**Expected Behavior**
All appropriate notices are generated/served.

**Actual Behavior**
The operation can time out if volume is high enough.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Cause of Bug, If Known**

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [ ] Deployed to the Court's `test` environment if prod-like data is required. Otherwise, deployed to any `experimental` environment for review.

**Comment by ttlenard** [comment]
@codyseibert @nechama-krigsman 

Some testing feedback:

1- The updating Contact info progress bar need a small grammatical update.  "a while" should be "awhile".
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBdmhEIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--3a9fc1ef5cf08f6292e77f85bbf14ff2e40a3e9f/image.png)

2 - I notice in the service logs that we are getting a lot of account throttled errors with service emails, and we  do retry to send the service emails.  Have we received any DLQ alerts on Test today with these?  One thing that I notice is that we are sending out the emails all at once (I see them all getting sent out around the same timestamp), which is most likely why we are getting the account throttled errors:
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBZ0ZFIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--0a7c212bcc270b1af3050962fcff827a51578d68/image.png)

3 - I noticed that when I changed an IRS Practitioners contact info, it took over 15 minutes to complete the request, but it eventually finished (not entirely sure if it updated all/sent all emails?).  I noticed in the logs that I got this error message. Cody mentioned that this error may have to do with the locking mechanism?  Do we need to do anything about this?
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBZ1JFIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--3e0d39142225cb4d30c0a7003a1513c9b5c5cd16/image.png)

**Comment by ttlenard** [comment]
Follow up with some more info about the locking mechanism stuff.

I had added a petitioner to case 32210-21 and then added in an email address that I have access to, just to check service emails.  I logged out of that case as the petitioner, then logged in as the IRS Practitioner, then changed my contact info.  I notice that in the logs, I see that the case that I had added the petitioner to was what threw the "entity is currently locked". I am stumped as to why this case was locked?  I wasn't logged in there as the petitioner when I updated the IRS Practitioners contact info.  Maybe @mmarcotte has some ideas?

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBazVFIiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--be7cf29e26af86f43ff16c0d9a4aac9859d2d515/image.png)
